snmp/datadog_checks/snmp/snmp.py

```diff
@@ -144,138 +144,175 @@ def discover_instances(self):
-            self._error = message
             raise CheckException(message)
 
-    def check_table(self, config, oids, lookup_names, enforce_constraints):
+    def check_table(self, config, table_oids):
         """
         Perform a snmpwalk on the domain specified by the oids, on the device
         configured in instance.
-        lookup_names is a boolean to specify whether or not to use the mibs to
-        resolve the name and values.
 
         Returns a dictionary:
         dict[oid/metric_name][row index] = value
         In case of scalar objects, the row index is just 0
         """
+        results = defaultdict(dict)
+        enforce_constraints = config.enforce_constraints
+        oids = []
+        bulk_oids = []
+        # Use bulk for SNMP version > 1 and there are enough symbols
+        bulk_limit = config.bulk_threshold if config.auth_data.mpModel else 0
+        for table, symbols in table_oids.items():
+            if not symbols:
+                # No table to browse, just one symbol
+                oids.append(table)
+            elif len(symbols) < bulk_limit:
+                oids.extend(symbols)
+            else:
+                bulk_oids.append(table)
+
+        all_binds, error = self.fetch_oids(config, oids, enforce_constraints=enforce_constraints)
+
+        for oid in bulk_oids:
+            try:
+                self.log.debug('Running SNMP command getBulk on OID %s', oid)
+                iterator = config.call_cmd(
+                    hlapi.bulkCmd,
+                    self._NON_REPEATERS,
+                    self._MAX_REPETITIONS,
+                    oid,
+                    lookupMib=enforce_constraints,
+                    ignoreNonIncreasingOid=self.ignore_nonincreasing_oid,
+                    lexicographicMode=False,
+                )
+                binds, error = self._consume_iterator(iterator, config)
+                all_binds.extend(binds)
+
+            except PySnmpError as e:
+                message = 'Failed to collect some metrics: {}'.format(e)
+                if not error:
+                    error = message
+                self.warning(message)
+
+        for result_oid, value in all_binds:
+            if not enforce_constraints:
+                # if enforce_constraints is false, then MIB resolution has not been done yet
+                # so we need to do it manually. We have to specify the mibs that we will need
+                # to resolve the name.
+                oid_to_resolve = hlapi.ObjectIdentity(result_oid.asTuple()).loadMibs(*config.mibs_to_load)
+                result_oid = oid_to_resolve.resolveWithMib(config.mib_view_controller)
+            _, metric, indexes = result_oid.getMibSymbol()
+            results[metric][indexes] = value
+        self.log.debug('Raw results: %s', results)
+        # Freeze the result
+        results.default_factory = None
+        return results, error
```

Comment:
It makes more sense this way. Note that the caller doesn't really modify the object, but it uses `get` on the result, which on a `defaultdict` is confusing, because the key is present but without a useful value.